<a href="https://colab.research.google.com/github/whoviren/whoviren/blob/main/EAI6010_VeerendraParshetty_Week3_FallB_2022.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# TITLE
**EAI6010 - Module 3: NLP AI Applications** 

**Submitted to - Professor Vladimir Shapiro**

# INTRODUCTION
Natural Language Processing (NLP) is a part of AI (artificial intelligence) that deals with understanding and processing of human language. NLP use casea are speech recognition, languague processing and sentimental analysis. In this project we will be using to Gutenberg corpus to analyze text.

# ANALYSIS

**Q1:** 
**A. Download and install the Gutenberg corpus tool to your Jupyter Notebook. Project**

Gutenberg contains some 25,000 free electronic books hosted at 
http://www.gutenberg.org/. We can install the NLTK package, then use the 
Gutenberg corpus in it. 

In [ ]:
!pip install nltk

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


**B. Download the Gutenberg corpus tool in the NLTK package.**

In [ ]:
import nltk
nltk.download('gutenberg')

[nltk_data] Downloading package gutenberg to /root/nltk_data...
[nltk_data]   Package gutenberg is already up-to-date!


True

**C. Use the texts in the corpus.**

Gutenberg electronic text archive contains some 25,000 free electronic books. We can use the file identifier command to check the files in this corpus

In [ ]:
nltk.corpus.gutenberg.fileids()

['austen-emma.txt',
 'austen-persuasion.txt',
 'austen-sense.txt',
 'bible-kjv.txt',
 'blake-poems.txt',
 'bryant-stories.txt',
 'burgess-busterbrown.txt',
 'carroll-alice.txt',
 'chesterton-ball.txt',
 'chesterton-brown.txt',
 'chesterton-thursday.txt',
 'edgeworth-parents.txt',
 'melville-moby_dick.txt',
 'milton-paradise.txt',
 'shakespeare-caesar.txt',
 'shakespeare-hamlet.txt',
 'shakespeare-macbeth.txt',
 'whitman-leaves.txt']

**D.  Create a table displaying relative frequencies with which “modals” (can, could, may, might, will, would, and should) are used in all texts provided in the corpus.**

In [ ]:
from nltk.corpus import gutenberg
modals = ['can', 'could', 'may', 'might', 'will', 'would','should']
cfdist = nltk.ConditionalFreqDist((word, file) 
                                  for file in gutenberg.fileids() 
                                  for word in gutenberg.words(file) 
                                  if word in modals)

In [ ]:
import pandas as pd
cfdist = pd.DataFrame(cfdist)
cfdist

would  could  might  will   may  should  can
austen-emma.txt            815    825    322   559   213     366  270
austen-persuasion.txt      351    444    166   162    87     185  100
austen-sense.txt           507    568    215   354   169     228  206
bible-kjv.txt              443    165    475  3807  1024     768  213
blake-poems.txt              3      3      2     3     5       6   20
bryant-stories.txt         110    154     23   144    18      38   75
burgess-busterbrown.txt     46     56     17    19     3      13   23
carroll-alice.txt           70     73     28    24    11      27   57
chesterton-ball.txt        139    117     69   198    90      75  131
chesterton-brown.txt       132    170     71   111    47      56  126
chesterton-thursday.txt    116    148     71   109    56      54  117
edgeworth-parents.txt      503    420    127   517   160     271  340
melville-moby_dick.txt     421    215    183   379   230     181  220
milton-paradise.txt         49     62     98   161   116      55  107
shakespeare-caesar.txt      40     18     12   129    35      38   16
shakespeare-hamlet.txt      60     26     28   131    56      52   33
shakespeare-macbeth.txt     42     15      5    62    30      41   21
whitman-leaves.txt          85     49     26   261    85      42   88

**E. For two modals with the largest span of relative frequencies (most used minus least used), select a text which uses it the most and the text that uses it the least. Compare usage in both texts by examining the relative frequencies of those modals in the two texts. Try to explain why those words are used differently in the two texts**. 

In [ ]:
# Adding up all the models used in corpus 
modal_count = cfdist.sum().sort_values(ascending=False)
modal_count

will      7130
would     3932
could     3528
should    2496
may       2435
can       2163
might     1938
dtype: int64

We now know that the most used model is the text bible and least used text is in the text blake. We should now find the 

In [ ]:
# Displying the max of modal 'will' in the corpus 
most_frequent_modal = modal_count.index[0]
print("Most Frequent Modal : "+ str(most_frequent_modal))
print("\n")
print("Text with maximum count of '" + str(most_frequent_modal) + "': " + cfdist.loc[cfdist[most_frequent_modal]==max(cfdist[most_frequent_modal]),most_frequent_modal].index[0])
print("\n")
print("Text with minimum count of '" + str(most_frequent_modal) + "': " + cfdist.loc[cfdist[most_frequent_modal]==min(cfdist[most_frequent_modal]),most_frequent_modal].index[0])

Most Frequent Modal : will


Text with maximum count of 'will': bible-kjv.txt


Text with minimum count of 'will': blake-poems.txt


In [ ]:
least_frequent_modal = modal_count.index[-1]
print("Least Frequent Modal : "+ str(least_frequent_modal))
print("\n")
print("Text with maximum count of '" + str(least_frequent_modal) + "': " + cfdist.loc[cfdist[least_frequent_modal]==max(cfdist[least_frequent_modal]),least_frequent_modal].index[0])
print("\n")
print("Text with minimum count of '" + str(least_frequent_modal) + "': " + cfdist.loc[cfdist[least_frequent_modal]==min(cfdist[least_frequent_modal]),least_frequent_modal].index[0])

Least Frequent Modal : might


Text with maximum count of 'might': bible-kjv.txt


Text with minimum count of 'might': blake-poems.txt


In [ ]:
text_bible =  nltk.Text(nltk.corpus.gutenberg.words('bible-kjv.txt'))
text_blake =  nltk.Text(nltk.corpus.gutenberg.words('blake-poems.txt'))

In [ ]:
# the concordance function shows us every occurrence of a given word

print(text_bible.concordance(most_frequent_modal))
print("\n")
print(text_blake.concordance(most_frequent_modal))

Displaying 25 of 3836 matches:
ood that the man should be alone ; I will make him an help meet for him . 2 : 
 the days of thy life : 3 : 15 And I will put enmity between thee and the woma
 . 3 : 16 Unto the woman he said , I will greatly multiply thy sorrow and thy 
 heart . 6 : 7 And the LORD said , I will destroy man whom I have created from
ence through them ; and , behold , I will destroy them with the earth . 6 : 14
rth shall die . 6 : 18 But with thee will I establish my covenant ; and thou s
h . 7 : 4 For yet seven days , and I will cause it to rain upon the earth fort
ry living substance that I have made will I destroy from off the face of the e
; and the LORD said in his heart , I will not again curse the ground any more 
art is evil from his youth ; neither will I again smite any more every thing l
 And surely your blood of your lives will I require ; at the hand of every bea
require ; at the hand of every beast will I require it , and at the hand of ma
at the hand of every 

In [ ]:
print(text_bible.concordance(least_frequent_modal))
print("\n")
print(text_blake.concordance(least_frequent_modal))

Displaying 25 of 475 matches:
aidst thou , She is my sister ? so I might have taken her to me to wife : now t
as not able to bear them , that they might dwell together : for their substance
raham said unto God , O that Ishmael might live before thee ! 17 : 19 And God s
ast done unto us ? one of the people might lightly have lien with thy wife , an
And Laban said , Behold , I would it might be according to thy word . 30 : 35 A
he cattle in the gutters , that they might conceive among the rods . 30 : 42 Bu
 me ; and didst not tell me , that I might have sent thee away with mirth , and
heir riches were more than that they might dwell together ; and the land wherei
, and lay no hand upon him ; that he might rid him out of their hands , to deli
y themselves : because the Egyptians might not eat bread with the Hebrews ; for
 Reuben , thou art my firstborn , my might , and the beginning of my strength ,
d the heart of his servants , that I might shew these my signs before him : 10 
e urgent u

**Q2:** 
**A. In the Inaugural corpus, see below:** 

In [ ]:
# Here we are loading the inagural corpus 
nltk.download('inaugural')
from nltk.corpus import inaugural

[nltk_data] Downloading package inaugural to /root/nltk_data...
[nltk_data]   Package inaugural is already up-to-date!


**B.Chose Kennedy’s speech, using e.g., this code:**

In [ ]:
# Choosing the '1961-Kennedy text' 
nltk.corpus.inaugural.words('1961-Kennedy.txt')

['Vice', 'President', 'Johnson', ',', 'Mr', '.', ...]

**C. Identify the 10 most frequently used long words (words longer than 7 characters).**

In [ ]:
# First we load all the words 

ken_words = nltk.corpus.inaugural.words('1961-Kennedy.txt')

In [ ]:
# Checking lenght of words
len(ken_words)

1546

In [ ]:
# Filtering the words lonnger than 7 charecters in Kennedy text and turning the words to lower case 

filt_char_7 = ([w.lower() for w in ken_words if len(w) > 7])

We can observe from the above table that the word 'citizens' had the highest occurences.

In [ ]:
# Frequency distribution

from nltk import FreqDist
f_dist = FreqDist(filt_char_7)
freq = f_dist.most_common(10)

In [ ]:
freq

[('citizens', 5),
 ('president', 4),
 ('americans', 4),
 ('generation', 3),
 ('forebears', 2),
 ('revolution', 2),
 ('committed', 2),
 ('powerful', 2),
 ('supporting', 2),
 ('themselves', 2)]

In [ ]:
# Creating an empty arrays
ken_1 = []
ken_2 = []

# Seperating the words
for i in range(10):
    ken_1.append(freq[i][0])

for i in range(10):
    ken_2.append(freq[i][1])

import pandas as pd

# Creating dataframe
df_freq_kennedy = pd.DataFrame({'Top 10':ken_1, 'Frequency':ken_2})

In [ ]:
# Printing top 10 words in Kennedy text with most occurences 
df_freq_kennedy

Top 10  Frequency
0    citizens          5
1   president          4
2   americans          4
3  generation          3
4   forebears          2
5  revolution          2
6   committed          2
7    powerful          2
8  supporting          2
9  themselves          2

In [ ]:
# Creating a list 
top_10_list = df_freq_kennedy['Top 10'].to_list()

In [ ]:
top_10_list

['citizens',
 'president',
 'americans',
 'generation',
 'forebears',
 'revolution',
 'committed',
 'powerful',
 'supporting',
 'themselves']

**D.  Which one of those 10 words has the largest number of synonyms? Use WordNet as a helper.**

In [ ]:
from nltk.corpus import wordnet as wn
nltk.download('wordnet')
nltk.download('omw-1.4')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


True

In [ ]:
# blank dictionary
synonyms = []

# traverse through top 10 words list created earlier 
for word in top_10_list:
  print(word)
  for j in wn.synsets(word):
    for s in j.lemmas():
      synonyms.append(s.name())
  print(set(synonyms))

citizens
{'citizen'}
president
{'United_States_President', 'president', 'President_of_the_United_States', 'President', 'chairman', 'chair', 'citizen', 'Chief_Executive', 'chairwoman', 'chairperson', 'prexy'}
americans
{'United_States_President', 'president', 'President_of_the_United_States', 'President', 'chairman', 'chair', 'American', 'American_English', 'citizen', 'Chief_Executive', 'chairwoman', 'chairperson', 'American_language', 'prexy'}
generation
{'President_of_the_United_States', 'Chief_Executive', 'chairperson', 'coevals', 'genesis', 'President', 'propagation', 'American_English', 'American_language', 'United_States_President', 'president', 'contemporaries', 'chair', 'citizen', 'prexy', 'chairman', 'American', 'chairwoman', 'multiplication', 'generation'}
forebears
{'President_of_the_United_States', 'Chief_Executive', 'chairperson', 'coevals', 'genesis', 'President', 'propagation', 'American_English', 'American_language', 'forebear', 'United_States_President', 'president', 'c

In [ ]:
# Creaty empty arrays
np_1 = []
np_2 = []

# Get synonyms
for i in range(10):
    np_1.append(np_syn[i][0])

for j in range(10):
    np_2.append(np_syn[j][1])

# Create dataframe
df_syn = pd.DataFrame({'Words':np_1, 'n':np_2})

In [ ]:
# Listing the word with largest number of synonyms
df_syn.sort_values(by="n", ascending=False)

Words   n
8  supporting  52
6   committed  27
1   president  16
7    powerful  16
3  generation  12
2   americans   5
5  revolution   5
4   forebears   2
0    citizens   1
9  themselves   0

**E. List all synonyms for the 10 most frequently used words. Which one of those 10 words has the largest number of hyponyms?** 

In [ ]:
# Create array
np_hyp = []

# Loop through most frequent distribution 
for i in f_dist.most_common(10):
    # Print word
    print ('Word: '), i[0]
    count=0
    
    # Print Hyponyms
    print ('Hyponyms: ')
    for j in wn.synsets(i[0]):
        print (j)
        print (j.hyponyms())
        count+= len(j.hyponyms())
    np_hyp.append([i[0], count])

Word: 
Hyponyms: 
Synset('citizen.n.01')
[Synset('active_citizen.n.01'), Synset('civilian.n.01'), Synset('freeman.n.01'), Synset('private_citizen.n.01'), Synset('repatriate.n.01'), Synset('thane.n.02'), Synset('voter.n.01')]
Word: 
Hyponyms: 
Synset('president.n.01')
[]
Synset('president_of_the_united_states.n.01')
[]
Synset('president.n.03')
[Synset('ex-president.n.01')]
Synset('president.n.04')
[Synset('kalon_tripa.n.01'), Synset('vice_chairman.n.01')]
Synset('president.n.05')
[]
Synset('president_of_the_united_states.n.02')
[]
Word: 
Hyponyms: 
Synset('american.n.01')
[Synset('african-american.n.01'), Synset('alabaman.n.01'), Synset('alaskan.n.01'), Synset('anglo-american.n.01'), Synset('appalachian.n.01'), Synset('arizonan.n.01'), Synset('arkansan.n.01'), Synset('asian_american.n.01'), Synset('bay_stater.n.01'), Synset('bostonian.n.01'), Synset('californian.n.01'), Synset('carolinian.n.01'), Synset('coloradan.n.01'), Synset('connecticuter.n.01'), Synset('creole.n.02'), Synset('dela

In [ ]:
# Creating empty arrays
np_1 = []
np_2 = []

# Get synonyms
for i in range(10):
    np_1.append(np_hyp[i][0])

for j in range(10):
    np_2.append(np_hyp[j][1])

# Create dataframe
df_hyp = pd.DataFrame({'Top 3':np_1, 'n':np_2})

**F. List all hyponyms of the 10 most frequently used words.**

In [ ]:
# Listing the words with most appearences
df_hyp.sort_values(by="n", ascending=False)

Top 3   n
2   americans  75
8  supporting  42
6   committed  16
5  revolution   8
0    citizens   7
3  generation   6
1   president   3
4   forebears   2
7    powerful   0
9  themselves   0

**G. Reflect on the results.**


*   In the Gutenber corpus we found out that will and might were the most and least frequently used words. Will was used 3807 in bible.text while might was 2 times in the same text.
*   In the same corpus 'might' was the least used modal and had the maximum count of 475 times in bible-kjv.txt, whereas had a minimum count of 2 in blake-poems.txt
*   The word 'supporting' had the highest number of sysnonyms in the Kennedy text with a count of 52.
*   In terms of hyponumns 'america' had the largest count with 75 occurences in the same text. 

**H.  Run all the cells, ensure all are executed and create the output. It’s your choice whether to execute locally or on Google Colab.**

Please hit Ctrl + F9 to run all cells 

# References

*   Hagmann, T. (n.d.). HW 8: Language processing. HW8_tim_hagmann. Retrieved November 19, 2022, from https://timhagmann.com/html/e63/hw8-hagmann-tim.html 
*   Accessing Text Corpora and Lexical Resources. 2. accessing text corpora and lexical resources. (n.d.). Retrieved November 19, 2022, from https://www.nltk.org/book/ch02.html 
*   Durai, K. (2022, May 31). Basics of Natural Language Processing for beginners. Medium. Retrieved November 19, 2022, from https://medium.com/geekculture/basics-of-natural-language-processing-for-beginners-d86351df9d09 





